internal/controllers/core/dockercomposelogstream/reconciler.go

```diff
@@ -2,48 +2,251 @@ package dockercomposelogstream
+func NewReconciler(client ctrlclient.Client, store store.RStore,
+	dcc dockercompose.DockerComposeClient, dc docker.Client) *Reconciler {
 	return &Reconciler{
-		client: client,
-		store:  store,
+		client:         client,
+		store:          store,
+		dcc:            dcc,
+		dc:             dc.ForOrchestrator(model.OrchestratorDC),
+		projectWatches: make(map[string]*ProjectWatch),
+		results:        make(map[types.NamespacedName]*Result),
+		containers:     make(map[serviceKey]*v1alpha1.DockerContainerState),
+		requeuer:       indexer.NewRequeuer(),
 	}
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	nn := req.NamespacedName
 	obj := &v1alpha1.DockerComposeLogStream{}
-	err := r.client.Get(ctx, req.NamespacedName, obj)
+	err := r.client.Get(ctx, nn, obj)
 	if err != nil && !apierrors.IsNotFound(err) {
 		return ctrl.Result{}, err
 	}
 
 	if apierrors.IsNotFound(err) || obj.ObjectMeta.DeletionTimestamp != nil {
+		r.clearResult(nn)
+		r.manageOwnedProjectWatches(ctx)
 		return ctrl.Result{}, nil
 	}
 
+	ctx = store.MustObjectLogHandler(ctx, r.store, obj)
+	r.manageOwnedProjectWatches(ctx)
+	r.manageLogWatch(ctx, nn, obj)
+
 	return ctrl.Result{}, nil
 }
 
+// Removes all state for an object.
+func (r *Reconciler) clearResult(nn types.NamespacedName) {
+	result, ok := r.results[nn]
+	if ok {
+		if result.watch != nil {
+			result.watch.cancel()
+		}
+		delete(r.results, nn)
+	}
+}
+
+// Looks up the container state for the current object, if possible.
+func (r *Reconciler) reconcileContainerState(ctx context.Context, obj *v1alpha1.DockerComposeLogStream, serviceKey serviceKey) {
+	id, err := r.dcc.ContainerID(ctx, v1alpha1.DockerComposeServiceSpec{Project: obj.Spec.Project, Service: obj.Spec.Service})
+	if err != nil {
+		return
+	}
+
+	state, err := r.getContainerState(ctx, string(id))
+	if err != nil {
+		return
+	}
+	r.recordContainerState(serviceKey, state)
+}
+
+// Starts the log watcher if necessary.
+func (r *Reconciler) manageLogWatch(ctx context.Context, nn types.NamespacedName, obj *v1alpha1.DockerComposeLogStream) {
+	// Make sure the result is up to date.
+	result, ok := r.results[nn]
+	changed := ok && !apicmp.DeepEqual(result.spec, obj.Spec)
+	if changed && result.watch != nil {
+		result.watch.cancel()
+		result.watch = nil
+	}
+
+	if !ok {
+		result = &Result{
+			name: nn,
+		}
+		r.results[nn] = result
+	}
+
+	if changed || !ok {
+		result.spec = obj.Spec
+		result.projectHash = dockercomposeservices.MustHashProject(obj.Spec.Project)
+	}
+
+	serviceKey := result.serviceKey()
+	r.reconcileContainerState(ctx, obj, serviceKey)
+
+	containerState := r.containers[serviceKey]
+	if containerState == nil || containerState.StartedAt.IsZero() {
+		// wait for the container to start before attaching so that we can filter out old logs
+		// for job containers that can be re-used
+		return
```

Comment:
Should this result in a `RequeueAfter` to make sure we come back and start watching logs once it's started? (Or maybe it's taken care of by the event watcher triggering a reconcile when it sees a container start?)

Reply:
ya, that's what the event-based requeuer is for!